guides/keras_core/migration_guide_tf_keras_3_0.py

```diff
@@ -0,0 +1,877 @@
+
+```python
+class MyCustomLayer(keras.layers.Layer):
+    def __init__(self, units):
+        super().__init__()
+        self.units = units
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self.w = tf.Variable(initial_value=tf.zeros([input_dim, self.units]))
+        self.b = tf.Variable(initial_value=tf.zeros([self.units,]))
+
+    def call(self, inputs):
+        return keras.ops.matmul(inputs, self.w) + self.b
+
+
+layer = MyCustomLayer(3)
+data = np.random.uniform(size=[3, 3])
+model = keras.models.Sequential([layer])
+model.compile(optimizer="adam", loss="mse")
+model.fit(data, data)
+# The model does not have any trainable variables
+for layer in model.layers:
+    print(layer.trainable_variables)
+```
+
+`UserWarning: The model does not have any trainable weights.
+  warnings.warn("The model does not have any trainable weights.")`
+"""
+
+"""
+Here is how you fix it:
+
+use `self.add_weight()` method or use a `keras.Variable` instead.
+
+"""
+
+
+class MyCustomLayer(keras.layers.Layer):
+    def __init__(self, units):
+        super().__init__()
+        self.units = units
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self.w = self.add_weight(
+            shape=[input_dim, self.units],
+            initializer="zeros",
+        )
+        self.b = self.add_weight(
+            shape=[
+                self.units,
+            ],
+            initializer="zeros",
+        )
+
+    def call(self, inputs):
+        return keras.ops.matmul(inputs, self.w) + self.b
+
+
+layer = MyCustomLayer(3)
+data = np.random.uniform(size=[3, 3])
+model = keras.models.Sequential([layer])
+model.compile(optimizer="adam", loss="mse")
+model.fit(data, data)
+# Verify that the variables are now being tracked
+for layer in model.layers:
+    print(layer.trainable_variables)
+
+"""
+### None entries are not allowed as part of nested (e.g. list/tuples) tensor arguments in
+Layer.call(), nor as part of call() return values.
+
+The following snippet of code will reproduce the error.
+
+```python
+class CustomLayer(keras.layers.Layer):
+    def __init__(self):
+        super().__init__()
+
+    def call(self, inputs):
+      return  inputs["foo"]
+
+layer = CustomLayer()
+inputs = {
+    "foo": keras.Input(shape=(1,), name="foo"),
+    "bar": {
+        "baz": None,
+    },
+}
+layer(inputs)
+```
+"""
+
+"""
+Here is how you will fix it:
+
+Replace `None` with a value or Keras Tensor
+"""
+
+
+class CustomLayer(keras.layers.Layer):
+    def __init__(self):
+        super().__init__()
+
+    def call(self, inputs):
+        output = inputs["foo"]
+        return None
+
+
+layer = CustomLayer()
+inputs = {
+    "foo": keras.Input(shape=(1,), name="foo"),
+    "bar": {
+        "baz": keras.Input(shape=(1,), name="bar"),
+    },
+}
+layer(inputs)
+
+"""
+### Random seed behavior changes between Keras 2 and Keras 3.
```

Comment:
Rather than mentioning the random seed, please show how to write a random layer in Keras 3, which is more useful for end users. You can say:

- Keras 3 has a new `keras.random` namespace with random ops
- All Keras 3 random ops are stateless: if seeded they return the same value every time (not so for TF random ops!!)
- Demo their behavior, contrasting it with TF random ops behavior
- To write a layer using these ops, you need to pass a different `seed` argument at each call. To do this, create a `SeedGenerator` instance in `__init__` then pass it as the `seed` argument in the random op in `call`.
- Demo a layer doing this (e.g. custom dropout)

Reply:
Updated the example. However I did not add the `seedGenerator` because adding a TF equivalent `tf.random.Generator` would make the random ops stateless in TF and I cannot demo the contrasting behavior. So I am just passing the seed.